components/credentials/src/main/kotlin/org/cloudfoundry/credhub/credential/CertificateCredentialValue.kt

```diff
@@ -55,6 +55,9 @@ class CertificateCredentialValue : CredentialValue {
     @JsonIgnore
     var versionCreatedAt: Instant? = null
 
+    @JsonInclude(Include.NON_NULL)
+    var durationOverridden: Boolean? = null
```

Comment:
Are you making this nullable for backwards compatibility with older certificates that won't have this property?

Reply:
The general intent behind it being nullable (here and in other spots) is so that when the feature is disabled, the API does not return the field at all. This allows the corresponding CLI update (see https://github.com/cloudfoundry-incubator/credhub-cli/pull/110) to:

1. Not display `duration_overridden` in CLI output when the server-side feature is disabled
2. Not display `duration_overridden` in CLI output when talking to older versions of the CredHub server that doesn't have this feature (perhaps not a large problem)
3. Display `duration_overridden: true` in CLI output when the server-side feature is enabled and overridden.

It seemed heavy handed to always display `duration_overridden` in CLI output, so we arrived at making this nullable. However, your other feedback about this field never being set to `false` is a good catch.  We're considering adding a 4th case:
4. Display `duration_overridden: false` in CLI output when the server-side feature is enabled but the duration is NOT overridden.

Reply:
OK thank you for the explanation, that makes a lot of sense.

There are 2 things I'd like to talk about: A general software point, and discussing these 4 options you mention.

I am a big believer in the [Software billion dollar mistake](https://hackernoon.com/null-the-billion-dollar-mistake-8t5z32d6). I believe that we should try to avoid using `null` as much as possible.
In this specific case, if you need a type containing 3 or 4 values, I would like to see some sort of enum created to represent all the possible options.

Regarding your specific question
>4. Display duration_overridden: false in CLI output when the server-side feature is enabled but the duration is NOT overridden.

I think that it does make sense.

I am also trying to think about the possibility of making the UX better. Putting myself in the shoes of a user, I am wondering how I feel about seeing a response containing `duration_overwritten: true`. "Overwritten to what?" I can hear myself think. Is the duration displayed in the CLI's output? If it is, then why bother add the "duration_overwritten" flag? The user will notice that it's not the one he asked for.

Completely different topic: forgive my non-native English, but I after some googling it seems to me like the proper spelling of the word is "overwritten". Google isn't too happy when you use "d"s instead of "t"s.

Reply:
We've updated the `duration_overridden` field to be non-nullable, so it will always appear regardless of whether the feature is enabled or not. That should make the API more consistent.

We have not renamed it - overridden is proper spelling, based off of the word override. Versus overwritten is based off of the word overwrote. Override seems more correct to us here, as the server is making a decision to reject the duration passed and use a different duration instead. But either word should work.

Reply:
>overridden is proper spelling, based off of the word override

Oh right, of course. I did some googling and I may have learned something about the different between "overwrite" and "override". Makes a lot of sense now, this appears to indeed be an override, and not an overwrite.

Reply:
> We've updated the duration_overridden field to be non-nullable, so it will always appear in the API response regardless of whether the feature is enabled or not. That should make the API more consistent.

Now we lose the possibility to disable the feature for old users who don't need it and don't want breaking changes. This is important to us, so that we can ship this in a patch release,  and not have to maintain multiple versions (as mentioned in another conversation).

My previous comments were not talking at the feature level, only about the implementation detail of this field. I think that I was pretty happy with the feature.

Here is my position, let me know if you disagree:
1. I think that we need a type that allows to represent 3 values:
    1. Duration overridden (and obviously "feature enabled") -- the previous `true`
    2. Feature enabled but duration kept -- the previous `false`
    3. Feature disabled -- the previous `null`
1. I really want to avoid `null`, so I'd rather not use a nullable boolean.

Hence, I think that we need an Enum type to represent all these values, and depending on the values, we can decide what we return back.